File: src/Java/algorithms/treeandgraph/BinaryTreeZigzagLevelOrderTraversal.java
package algorithms.treeandgraph; 

import java.util.*; 
import lib.TreeNode; 

public class BinaryTreeZigzagLevelOrderTraversal {
    /* Function: Traverse binary tree in level order and the even row is in reverse order. 
     * Solution: Collections have method reverse. 
     */ 
    
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>(); 
        if (root == null) {
            return res; 
        }
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>(); 
        queue.offer(root); 
        
        int lev = 0; 
        while (!queue.isEmpty()) {
            int size = queue.size(); 
            List<Integer> list = new ArrayList<Integer>(); 
            for (int i = 0; i < size; ++i) {
                TreeNode cur = queue.poll(); 
                list.add(cur.val); 
                if (cur.left != null) {
                    queue.offer(cur.left); 
                }
                if (cur.right != null) {
                    queue.offer(cur.right); 
                }
            }
            
            if ((lev++) % 2 == 0) {
                res.add(list); 
            } else {
                Collections.reverse(list); 
                res.add(list); 
            }
        }
        
        return res; 
    }
} 

File: src/CPP/src/leetcode/BinaryTreeZigzagLevelOrderTraversal.cpp
/*
 * [Source] https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/
 * [Difficulty]: Medium
 * [Tag]: Tree
 * [Tag]: Breadth-first Search
 * [Tag]: Stack
 */

#include <iostream>

using namespace std;

// [Solution]: Level order traverse with odd level lists reversed.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/BinaryTreeZigzagLevelOrderTraversal.java
 */

/* Java solution
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
		List<List<Integer>> result=new ArrayList<List<Integer>>();
		if(root==null) return result;
		levelOrder(root, 0, result);
		int i=1;
		while(i<result.size()){
			Collections.reverse(result.get(i));
			i=i+2;
		}
		return result;
    }
	
	public void levelOrder(TreeNode root, int level, List<List<Integer>> result){
		if(root==null) return;
		if(result.size()<=level){
			List<Integer> list=new ArrayList<Integer>();
			list.add(root.val);
			result.add(list);
		}else{
			result.get(level).add(root.val);
		}
		levelOrder(root.left, level+1, result);
		levelOrder(root.right, level+1, result);
	}
}
 */

int main() {
    Solution sol;

    return 0;
}
